### ISSUE TITLE
Uploading a new version of a file removes the old version from downloads before the upload is complete

### ISSUE THREAD

**Issue by slowriot** [leafo/itch.io#145]
In the case of large uploads, it can take us over an hour to upload new versions of binaries - during that time the existing version disappears from the download options.  We've just encountered a problem where a reviewer was unable to obtain a Windows version of our game because we were mid-update.

The ideal behaviour would be to only replace the original file once upload has successfully completed.

**Comment by Tectorum** [comment]
An alternative---deleting the file in the game edit menu and uploading a new file of the same name---likewise immediately deletes the file. I'd really like to see changes not take effect to the game until the "Save" button is clicked.

**Comment by Guard13007** [comment]
@fasterthanlime Does this issue affect builds pushed with butler, or no?

**Comment by fasterthanlime** [comment]
@Guard13007 nope, that's only for web uploads.

(I'm also unsure if that's still the current behavior)

**Comment by Guard13007** [comment]
Well, even though I switched to butler (and am loving it), I think this deserves some love, I'm guessing butler isn't being hugely used yet. (Although, I hope it is, cause.. it's awesome!)

**Comment by josefnpat** [comment]
I agree; using butler's bdiff feature saves me a ton of time and bandwidth. Perhaps allowing the web version of itch.io to accept bdiffs would be nice?

**Comment by fasterthanlime** [comment]
> I'm guessing butler isn't being hugely used yet.

Yes and no: I'm counting over 25K builds pushed with butler - but that's *still* a minority of itch.io uploads :)

> Perhaps allowing the web version of itch.io to accept bdiffs would be nice?

I've actually played with cross-compiling butler to javascript for the diff feature, but the HTML filesystem APIs and javascript execution speed killed the project. I think a good first step to making butler more accessible is allowing drag & drop via the app, that would use butler under the hood.

~~~

I also agree that the web upload replace behavior should be improved - I have even bigger gripes with it: when "updating" a game this way, an upload disappears, making way for a new one. This is an issue for itch, since it has to rely on approximate heuristics to choose the "new best upload" to update to. Sometimes things go wrong: https://github.com/itchio/itch/issues/1129

My personal dream timeline is one where all web upload get an auto-generated channel name and their files get processed by wharfd the same way butler builds do, which mean we get things for free like:

  * Re-compressing correctly
  * Fixing permissions
  * Patching
  * Scanning contents & automatically tagging platforms (soon)
  * Scanning contents for harmful stuff (soon)

There's a few things blocking this, [single-file support in wharfd](https://github.com/itchio/butler/issues/58) is one of them. It'll happen eventually though!
